src/java/org/apache/cassandra/index/sai/utils/RowIdToPrimaryKeyMapper.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.utils;
+
+import org.apache.cassandra.index.sai.disk.PostingList;
+import org.apache.cassandra.index.sai.disk.PrimaryKeyMap;
+
+/**
+ * Essentially a Function<Long, PrimaryKey>, but avoids the boxing of the long.
+ */
+public class RowIdToPrimaryKeyMapper
+{
+    private final PrimaryKeyMap primaryKeyMap;
+    private final long segmentRowIdOffset;
+
+    /**
+     *
+     * @param primaryKeyMap - primary key generator
+     * @param segmentRowIdOffset - the offset to add to the rowId to get the SS Table row id
+     */
+    public RowIdToPrimaryKeyMapper(PrimaryKeyMap primaryKeyMap, long segmentRowIdOffset)
+    {
+        this.primaryKeyMap = primaryKeyMap;
+        this.segmentRowIdOffset = segmentRowIdOffset;
+    }
+
+    public PrimaryKey getPrimaryKeyForRowId(long rowId)
+    {
+        long ssTableRowId = rowId + segmentRowIdOffset;
+        // Special case where we wouldn't want to attempt to retreive a PK
+        if (ssTableRowId == PostingList.END_OF_STREAM)
+            return null;
+
+        return primaryKeyMap.primaryKeyFromRowId(ssTableRowId);
```

Comment:
I am getting exceptions in this method due to buffer overflow. Looks like I have a concurrency issue given that the issues disappear when I add the `synchronized` keyword to the method. Looking into it more.

Reply:
@JeremiahDJordan - the problem is coming from the `.parallel()` call in this block of test code:

https://github.com/datastax/cassandra/blob/9cabec504d9f70ddfb345122dc4defed56a40f85/test/unit/org/apache/cassandra/index/sai/cql/VectorSiftSmallTest.java#L118-L141

I think the primary issue is that the call to `execute` is not actually issuing multiple calls, but is parallel calls on the same instances. Doing a bit more digging.

Reply:
Notably, when I port that `VectorSiftSmallTest` into the `VectorDistributedTest`, I am getting concurrency issues in cassandra as well, so the issues are likely from the changes in this PR. I'll keep digging tomorrow.

Reply:
`VectorSiftSmallTest` is now passing with the thread safety fixes.